systemtest/benchtest/expvarmetrics/metrics.go

```diff
@@ -0,0 +1,253 @@
+// refers to an error in the collector. Both cases end the
+// lifecycle of a watcher.
+func (c *Collector) AddWatch(m Metric, expected int64) (<-chan bool, error) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if _, ok := c.watchers[m]; ok {
+		return nil, errors.New("watcher already exists for the given metric")
+	}
+
+	out := make(chan bool)
+	c.watchers[m] = &watchItem{
+		notifyChan: out,
+		validator:  func(in int64) bool { return in == expected },
+	}
+
+	return out, nil
+}
+
+func (c *Collector) addExpvar(e expvar) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	c.processMetric(Goroutines, e.Goroutines)
+	c.processMetric(Bytes, e.UncompressedBytes)
+	c.processMetric(TotalEvents, e.TotalEvents)
+	c.processMetric(ActiveEvents, e.ActiveEvents)
+	c.processMetric(RSSMemoryBytes, e.RSSMemoryBytes)
+	c.processMetric(AvailableBulkRequests, e.AvailableBulkRequests)
+	c.processMetric(TransactionsProcessed, e.TransactionsProcessed)
+	c.processMetric(SpansProcessed, e.SpansProcessed)
+	c.processMetric(MetricsProcessed, e.MetricsProcessed)
+	c.processMetric(ErrorsProcessed, e.ErrorsProcessed)
+	c.processMetric(ErrorElasticResponses, e.ErrorElasticResponses)
+	c.processMetric(ErrorOTLPTracesResponses, e.ErrorOTLPTracesResponses)
+	c.processMetric(ErrorOTLPMetricsResponses, e.ErrorOTLPMetricsResponses)
+	c.processMetric(NumGC, int64(e.NumGC))
+	c.processMetric(PauseTotalNs, int64(e.PauseTotalNs))
+	c.processMetric(MemAllocs, int64(e.Mallocs))
+	c.processMetric(MemBytes, int64(e.TotalAlloc))
+	c.processMetric(HeapAlloc, int64(e.HeapAlloc))
+	c.processMetric(HeapObjects, int64(e.HeapObjects))
+}
+
+func (c *Collector) processMetric(m Metric, val int64) {
+	if _, ok := c.metrics[m]; !ok {
+		c.metrics[m] = &AggregateStats{}
+	}
+	c.updateMetric(c.metrics[m], val)
+
+	if watcher, ok := c.watchers[m]; ok && watcher.validator(val) {
+		watcher.notifyChan <- true
+		delete(c.watchers, m)
+	}
+}
+
+func (c *Collector) updateMetric(stats *AggregateStats, value int64) {
+	// Initialize for first value
+	if stats.sampleCount == 0 {
+		stats.First = value
+		stats.Min = value
+		stats.Max = value
+	}
+
+	stats.Last = value
+	stats.sampleCount += 1
+	stats.Mean += (float64(value) - stats.Mean) / float64(stats.sampleCount)
+	if stats.Min > value {
+		stats.Min = value
+	}
+	if stats.Max < value {
+		stats.Max = value
+	}
+}
+
+func (c *Collector) rejectWatchers() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	for _, watcher := range c.watchers {
+		watcher.notifyChan <- false
+	}
+}
+
+func (c *Collector) start(ctx context.Context) error {
+	var first expvar
+	err := queryExpvar(ctx, &first, c.serverURL)
+	if err != nil {
+		return err
+	}
+	c.addExpvar(first)
+
+	outChan, errChan := run(ctx, c.serverURL, c.period)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				c.rejectWatchers()
+				return
+			case <-errChan:
+				c.rejectWatchers()
+				return
+			case m := <-outChan:
+				c.addExpvar(m)
+			}
+		}
+	}()
+	return nil
+}
```

Comment:
I think we can modify the signature to accept the `serverURL` and `period` to avoid storing them in the Collector structure, since those values aren't used elsewhere.

Also, the `rejectWatchers()` is called twice on error or context cancellation, we can defer calling that function inside the goroutine so the code doesn't have to be repeated.

```suggestion
func (c *Collector) start(ctx context.Context, serverURL string, period time.Duration) error {
	var first expvar
	if err := queryExpvar(ctx, &first, serverURL); err != nil {
		return err
	}
	c.addExpvar(first)

	outChan, errChan := run(ctx, serverURL, period)
	go func() {
		defer c.closeWatches()
		for {
			select {
			case <-ctx.Done():
				return
			case <-errChan:
				return
			case m := <-outChan:
				c.addExpvar(m)
			}
		}
	}()
	return nil
}
```

Reply:
ACK, looks cleaner